core/src/main/java/org/apache/servicecomb/core/SCBEngine.java

```diff
@@ -0,0 +1,272 @@
+  public void onInvocationStart(InvocationStartEvent event) {
+    invocationStartedCounter.incrementAndGet();
+  }
+
+  @AllowConcurrentEvents
+  @Subscribe
+  public void onInvocationFinish(InvocationFinishEvent event) {
+    invocationFinishedCounter.incrementAndGet();
+  }
+
+  public synchronized void init() {
+    if (SCBStatus.DOWN.equals(status)) {
+      try {
+        doInit();
+        status = SCBStatus.UP;
+      } catch (Exception e) {
+        uninit();
+        status = SCBStatus.FAILED;
+        throw new IllegalStateException("ServiceComb init failed.", e);
+      }
+    }
+  }
+
+
+  private void doInit() throws Exception {
+    status = SCBStatus.STARTING;
+
+    eventBus.register(this);
+
+    AbstractEndpointsCache.init(RegistryUtils.getInstanceCacheManager(), transportManager);
+
+    triggerEvent(EventType.BEFORE_HANDLER);
+    HandlerConfigUtils.init();
+    triggerEvent(EventType.AFTER_HANDLER);
+
+    triggerEvent(EventType.BEFORE_PRODUCER_PROVIDER);
+    producerProviderManager.init();
+    triggerEvent(EventType.AFTER_PRODUCER_PROVIDER);
+
+    triggerEvent(EventType.BEFORE_CONSUMER_PROVIDER);
+    consumerProviderManager.init();
+    triggerEvent(EventType.AFTER_CONSUMER_PROVIDER);
+
+    triggerEvent(EventType.BEFORE_TRANSPORT);
+    transportManager.init();
+    triggerEvent(EventType.AFTER_TRANSPORT);
+
+    schemaListenerManager.notifySchemaListener();
+
+    triggerEvent(EventType.BEFORE_REGISTRY);
+
+    triggerAfterRegistryEvent();
+
+    RegistryUtils.run();
+
+    Runtime.getRuntime().addShutdownHook(new Thread(this::uninit));
+  }
+
+  /**
+   * not allow throw any exception
+   * even some step throw exception, must catch it and go on, otherwise shutdown process will be broken.
+   */
+  public synchronized void uninit() {
+    if (SCBStatus.UP.equals(status)) {
+      LOGGER.info("ServiceComb is closing now...");
+      try {
+        doUninit();
+        status = SCBStatus.DOWN;
+      } catch (Exception e) {
+        status = SCBStatus.FAILED;
+        LOGGER.info("ServiceComb failed shutdown", e);
+      }
+    }
+  }
+
+  private void doUninit() throws Exception {
+    //Step 1: notify all component stop invoke via BEFORE_CLOSE Event
+    safeTriggerEvent(EventType.BEFORE_CLOSE);
+
+    status = SCBStatus.STOPPING;
+
+    //Step 2: Unregister microservice instance from Service Center and close vertx
+    // Forbidden other consumers find me
+    RegistryUtils.destroy();
+    VertxUtils.blockCloseVertxByName("registry");
+
+    //Step 3: wait all invocation finished
+    // forbit create new consumer invocation
+    validAllInvocationFinished();
+
+    //Step 4: Stop vertx to prevent blocking exit
+    VertxUtils.blockCloseVertxByName("config-center");
+    VertxUtils.blockCloseVertxByName("transport");
+
+    //Step 5: notify all component do clean works via AFTER_CLOSE Event
+    safeTriggerEvent(EventType.AFTER_CLOSE);
+
+    //Step 6: Clean flags for re-init
+    eventBus.unregister(this);
```

Comment:
when succeed to init, then already unregister this, unregister a not register subscribe will cause exception.

Reply:
yes,it had unregister in triggerAfterRegistryEvent after succeed to init,I will delete this line